mountpoint-s3/src/fs.rs

```diff
@@ -139,20 +144,29 @@ impl<Client: ObjectClient> UploadState<Client> {
                             error!(?err, "error updating the inode status");
                         }
                     }
-                    Self::Failed(_) | Self::Completed => unreachable!("checked by get_upload_in_progress"),
+                    Self::Failed(_) | Self::Completed => unreachable!("checked above"),
                 };
                 Err(e.into())
             }
         }
     }
 
-    async fn complete(&mut self, key: &str) -> Result<(), Error> {
-        // Check that the upload is still in progress.
-        _ = self.get_upload_in_progress(key)?;
+    async fn complete(&mut self, key: &str, ignore_if_empty: bool) -> Result<(), Error> {
+        match self {
+            Self::InProgress { request, .. } => {
+                if ignore_if_empty && request.size() == 0 {
+                    return Ok(());
```

Comment:
Maybe log here to help our future selves debugging `flush`-related stuff:
```suggestion
                    trace!(key, "not completing upload because file is empty");
                    return Ok(());
```